chore: adding new `COPY` options to reference #953 (#955)
questdb/questdb.io: docs/reference/sql/copy.md

## Description

Copies data from a delimited text file into a table. `COPY` has the following import modes:

- If the original source table is not partitioned, serial import is used:

  - If the target table exists, then the data is appended provided the file
  structure matches the table.
  - If the target table does not exist, then it is created using metadata derived
  from the file data.

> 

Reviewer: source vs target table  ? 
I think it's better to use 'target table' . 

Logic is a bit more complicated because parallel import is used when :
- target table exists  and is partitioned 
- target table doesn't exist and PARTITION BY is specified (to a valid unit ).

- If the original source table is partitioned, parallel import is used:

  - If the target table exist, it must be empty for the operation to perform successfully.
  - If the target table does not exist, `TIMESTAMP` or `FORMAT` must be defined and `PARTITION BY` cannot be `NONE`, `null`, or skipped.

>   - The operation takes up all the available resources, and therefore only one parallel import is allowed at a time.

Reviewer: git ?

## Description

> Copies tables from a delimited text file into a table. `COPY` has the following import modes:

Reviewer 2: nit: "Copies tables into a table" sounds weird. The input file contains lines or rows and they're inserted into a single table.

Reviewer 2: Would be nice to address this one. @amyshwang do you think current text makes sense to the end user?

Author: I was thinking about it and forgot to make change! I was debating whether to write:
```
Imports tables from a delimited text file into a table
```

It makes more sense but is far away from the key word `COPY`.

I will try:
```
Copies tables from a delimited text file saved in the defined root directory into QuestDB.
```

- Parallel import, used for copying partitioned tables:

>   - If the target table exists and is partitioned, the target table must be empty.

Reviewer: FORMAT is not strictly necessary but should be used if file uses format other than - yyyy-MM-ddTHH:mm:ss.SSSUUUz .

Author: But if `TIMESTAMP` is absent and `FORMAT` is not `- yyyy-MM-ddTHH:mm:ss.SSSUUUz` then one of them needs to be defined, right?

Author: Updated but not sure if it is correct.

>   - If the target table does not exist, `TIMESTAMP` or `FORMAT` must be defined, and `PARTITION BY` should not be `NONE`, `null`, or skipped.

Reviewer 2: We don't support `null` value for PARTITION BY, so I'm not sure if we should mention this invalid value.

  ::: note

> 

Reviewer: Should we mention that while one import is running new request(s) will be rejected or show that in tutorial ?

Reviewer: The message returned by qdb is  :

"Another import request is in progress. [activeImportId=4650934181329334916]"

it's a bit inconsistent with the table so we should fix it .

Reviewer 2: > Should we mention that while one import is running new request(s) will be rejected or show that in tutorial ?

Yeah, would be nice to mention this.

Reviewer 2: > it's a bit inconsistent with the table so we should fix it .

I'll submit a patch.

Author: Adding it here.

>   Parallel import takes up all the available resources, and therefore only one parallel import is allowed at a time.

Reviewer 2: Both COPY implementations don't support concurrent imports, so this note is misleading.

Reviewer: ok, so that'sa difference compared to old implementation which allowed loading multiple tables in parallel (with multiple copy commands) .

  :::

- Serial import, used for copying non-partitioned tables:

  - If the target table exists and is not partitioned. The data is appended provided the file structure matches the table.
  - If the target table does not exist, then it is created using metadata derived
  from the file data.

### Log table

`COPY` generates a log table,`sys.text_import_log`, tracking `COPY` operation for the last three days with the following information:

| Column name   | Data type | Notes                                                           |
|---------------|-----------|-----------------------------------------------------------------|
| ts            | timestamp | The log event timestamp                                         |
| id            | string    | Import id                                                       |
| table         | symbol    | Destination table name                                          |
| file          | symbol    | The source csv file                                             |

> | phase         | symbol    | Import phase.*  Available only to parallel import               |

Reviewer 2: ```suggestion
| phase         | symbol    | Import phase.* Available only in intermediate log records of parallel import               |
```

| status        | symbol    | The event status: started, finished, failed, cancelled          |
| message       | string    | The error message for when status is failed                     |

> | rows_handled  | long      | The counters for the total number of scanned lines in the file  |

Reviewer 2: We should mention that this and the next counters are only available in the final log row for the given import.

| rows_imported | long      | The counters for the total number of imported rows              |
| errors        | long      | The number of errors for the given phase                        |

\* Available phases for parallel import are:

- setup
- boundary_check
- indexing
- partition_import
- symbol_table_merge
- update_symbol_keys
- build_symbol_index
- move_partitions
- attach_partitions
- analyze_file_structure
- cleanup

> Table message retention is configurable through `cairo.sql.copy.log.retention.days` setting, and is three days by default.

Reviewer 2: ```suggestion
Log table row retention is configurable through `cairo.sql.copy.log.retention.days` setting, and is three days by default.
```

> `COPY` returns `id` value from `sys.text_import_log` to facilitate debugging.

Reviewer 2: Why do we mention debugging here? We return the id to allow user to track the import progress and await until import finishes. It has nothing to do with debugging.

Reviewer: I think Amy meant troubleshooting but unless we add a whole section on it it's better to use progress tracking.

> :::note

Reviewer 2: nit: I'd move this and subsequent paragraph to the top of this section. This way it'd be more apparent for the user that they need to change the config to be able to user imports. Would be also nice to mention explicitly that they have to copy their CSV files into the root dir before running an import.

> :::note

Reviewer 2: We should also mention `cairo.sql.copy.work.root` directory in this section.

## Description

Copies tables from a delimited text file saved in the defined root directory into a table. `COPY` has the following import modes:

- Parallel import, used for copying partitioned tables:

  - If the target table exists and is partitioned, the target table must be empty.

>   - If the target table does not exist, and the timestamp format is not `yyyy-MM-ddTHH:mm:ss.SSSUUUz`, `TIMESTAMP` or `FORMAT` must be defined, and `PARTITION BY` should not be `NONE`, or skipped.

Reviewer 2: > `TIMESTAMP` or `FORMAT` must be defined

In fact, we only require both to be specified in the described scenario. I'd propose simplifying this sentence to:
```
If the target table does not exist, both `TIMESTAMP` and `PARTITION BY` options must be defined to create a partitioned table and use parallel import. The `PARTITION BY` value should not be `NONE`.
```

Author: Thanks - removing "use parallel import" from your suggestion as the bullet point talks about parallel import and we can make it even shorter.

- Serial import, used for copying non-partitioned tables:

>   - If the target table exists and is not partitioned. The data is appended provided the file structure matches the table.

Reviewer 2: Should be one sentence:
```suggestion
  - If the target table exists and is not partitioned, the data is appended provided the file structure matches the table.
```

  - If the target table does not exist, then it is created using metadata derived
  from the file data.

:::note

 `COPY` takes up all the available resources. While one import is running, new request(s) will be rejected.

:::

`COPY '<id>' CANCEL` cancels the copying operation defined by the import `id`, while an import is taking place. 

### Root directory

`COPY` requires a defined root directory where CSV files are saved and copied from. A CSV file must be saved to the root directory before starting the `COPY` operation. There are two root directories to be defined:

> - `cairo.sql.copy.work.root` is used for storing temporary files like indexes or temporary partitions.

Reviewer 2: We should mention that this property is not mandatory since it defaults to the same value as `cairo.sql.copy.root`. So, only `cairo.sql.copy.root` is a must-have for COPY to be functional.

## Options

- `WITH HEADER true/false`: When `true`, QuestDB automatically assumes the first row is a
  header. Otherwise, schema recognition is used to determine whether the first row is used as header.

> - `PARALLEL`: When enabled, parallel import is on. By default this option is disabled and sequential import is used.

Reviewer 2: We should also document that only a single parallel import is allowed to be executed, i.e. users can't run multiple imports to be executed concurrently.

Author: Also to add these thanks to @bziobrowski 

- Old sequential import is synchronous and returns (in web console or sql) once import is done 
- New import is async, it returns 'ok' message after receiving the command but then it queues it and runs in background
- only one such import can run at a time because it's supposed to use all available server resources

Reviewer 2: Yes and we need to document all of these aspects properly, including what's written into the `sys.parallel_text_import_log` table.

Reviewer 2: One more thing to add: we may decide that we want to align old and parallel COPY behavior (although, it'll be a breaking change) - ideally the exact implementation should be chosen automatically, so that the user doesn't have to specify `with parallel` flag; plus, both of them should become async (hence, write their status into the log table). Let me post an update here once the decision is made.

Author: @puzpuzpuz I added a new commit for things I understand but I don't understand the bit on `sys.parallel_text_import_log` :)

Reviewer 2: It's the table that accumulates all import status updates.

> Let me post an update here once the decision is made.

Yup, we'll be merging COPY implementations into a single one. So, let's wait until things get more stable and then update this PR. Does it make sense?

- `TIMESTAMP`: Define the name of the timestamp column in the file to be imported.
- `FORMAT`: Timestamp column format when the format is not standard or cannot be detected.
- `PARTITION BY`: Partition unit to use when target table does not exist.

> - `On ERROR`: Define responses to data parsing errors. The valid values are:

Reviewer 2: ```suggestion
- `ON ERROR`: Define responses to data parsing errors. The valid values are:
```

:::

The `copy root directory` can be on a local disk to the server, on a remote disk, or
a remote filesystem. QuestDB enforces that the tables are only written from
files located in a directory relative to the `copy root directory`. This is a
security feature preventing random file access by QuestDB.

## Options

- `WITH HEADER true/false`: When `true`, QuestDB automatically assumes the first row is a
  header. Otherwise, schema recognition is used to determine whether the first row is used as header.
- `TIMESTAMP`: Define the name of the timestamp column in the file to be imported. This is required for parallel import when the table does not exist.
- `FORMAT`: Timestamp column format when the format is not standard or cannot be detected.
- `PARTITION BY`: Partition unit. This is required for parallel import when the table does not exist, and the value cannot be `NONE` or `null`.

> - `ON ERROR`: Define responses to data parsing errors. The valid values are:

Reviewer: I think it'd make sense to add description to on error options, e.g. 
SKIP_ROW - skip entire row
SKIP_COLUMN - skip column and use default value (null for nullable types, false for boolean, 0 for other non-nullable types)
ABORT - abort whole import on first error, restoring pre-import table status .

:::

The `copy root directory` can be on a local disk to the server, on a remote disk, or
a remote filesystem. QuestDB enforces that the tables are only written from
files located in a directory relative to the `copy root directory`. This is a
security feature preventing random file access by QuestDB.

> ## Options

Reviewer 2: We should describe the default values for all options.

Reviewer 2: Here is the list of options with default values:
* `header` - `false`
* `format` - `'yyyy-MM-ddTHH:mm:ss.SSSUUUZ'`
* `delimiter` - `','`

> 

Reviewer: "WITH" is not part of HEADER option .
WITH behaves as a obligatory 'separator' between copy X from 'source.csv' and options . 
(in PostgreSQL it's optional but let's leave it as is) .

Reviewer 2: Same as https://github.com/questdb/questdb.io/pull/955#discussion_r938530841

> - `WITH HEADER true/false`: When `true`, QuestDB automatically assumes the first row is a

Reviewer 2: `WITH` keyword has to be placed before the very first option, but necessarily `HEADER true/false`.

  header. Otherwise, schema recognition is used to determine whether the first row is used as header.
- `TIMESTAMP`: Define the name of the timestamp column in the file to be imported.
- `FORMAT`: Timestamp column format when the format is not standard or cannot be detected.
- `PARTITION BY`: Partition unit.
- `ON ERROR`: Define responses to data parsing errors. The valid values are:
    - `SKIP_ROW`: Skip the entire row
    - `SKIP_COLUMN`: Skip column and use the default value (`null` for nullable types, `false` for boolean, `0` for other non-nullable types)
    - `ABORT`: Abort whole import on first error, and restore the pre-import table status

> - `CANCEL`: `COPY '<id>' CANCEL`: using the import `id` to cancel copying a table.

Reviewer 2: `CANCEL` is not an option, but another type of `COPY` statement.

## Examples

> For more details on parallel import, please also see [Importing data in bulk via CSV](/docs/guides/importing-data/).

Reviewer 2: The linked section doesn't stand for parallel import, but rather HTTP-based import. That one is serial, but we shouldn't mention this as it's an implementation detail. In general, we should mention HTTP-based import as a more convenient alternative to `COPY` for the case of smaller CSV files.

Author: We're adding parallel import to the page: https://github.com/questdb/questdb.io/pull/967

Reviewer 2: I didn't see that PR when writing this comment. Resolving it now.

```questdb-sql title="COPY"
COPY weather FROM 'weather.csv' WITH HEADER true FORMAT 'yyyy-MM-ddTHH:mm:ss.SSSUUUZ' ON ERROR SKIP_ROW;
```

> Returns a hexadecimal import id:

Reviewer 2: nit: the id format doesn't matter much and is a subject to change in future versions
```suggestion
Returns an import id string:
```

> Returns a hexadecimal import id:

Reviewer 2: nit: would be nice to say that the above SQL statement starts the import asynchronously and returns the import id string.

> |  id string      |

Reviewer 2: nit: no need to include the type here
```suggestion
|  id       |
```

|-----------------|
|55ca24e5ba328050 |

The log can be accessed by querying:

```questdb-sql

> SELECT * FROM 'sys.text_import_log'

Reviewer 2: nit: it makes more sense to show how to check logs of a concrete import
```suggestion
SELECT * FROM 'sys.text_import_log' WHERE id = '55ca24e5ba328050';
```

```

> A sample log table low:

Reviewer 2: nit: what does "low" mean?

|   ts                           |   id                |   table  |   file         |   phase  |   status   |   message  |   rows_handled  |   rows_imported  |   errors  |
|--------------------------------|---------------------|----------|----------------|----------|------------|------------|-----------------|------------------|-----------|
|   2022-08-03T10:40:25.586455Z  |   55ca24e5ba328050  |  weather |   weather.csv  |          |   started  |            |                 |                  |   0       |
|                                |                     |          |                |          |            |            |                 |                  |           |

While it's running, import can be cancelled with:

```questdb-sql

> copy '55ca24e5ba328050' cancel;

Reviewer 2: ```suggestion
COPY '55ca24e5ba328050' CANCEL;
```

```

Within a few seconds import should stop and message with 'cancelled' status should appear in text_import_log, e.g.  

```questdb-sql

> select * from sys.text_import_log where id = '55ca24e5ba328050' order by ts desc limit 1

Reviewer 2: ```suggestion
SELECT * FROM 'sys.text_import_log' WHERE id = '55ca24e5ba328050' LIMIT -1;
```
